### PR TITLE
Initial TypeDescriptor implementation

### PR DESCRIPTION
**Pull request by chlowell**
This is an initial buildable implementation of TypeDescriptor. Some code remains behind feature flags. Several tests must be skipped because ReflectTypeDescriptionProvider is implemented with stubs.

cc @twsouthwick

### CODE REVIEW

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptionProvider.cs [hunk 1]
```diff
@@ -218,12 +215,12 @@ public virtual Type GetRuntimeType(Type reflectionType)
                 throw new ArgumentNullException("reflectionType");
```
Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptionProvider.cs]: `nameof(reflectionType)`

Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptionProvider.cs]: `nameof(reflectionType)`

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs [hunk 2]
```diff
@@ -2665,7 +2662,8 @@ public static void SortDescriptorArray(IList infos)
         [Conditional("DEBUG")]
         internal static void Trace(string message, params object[] args)
         {
-            Debug.WriteLineIf(s_traceDescriptor.Enabled, string.Format(CultureInfo.InvariantCulture, "TypeDescriptor : {0}", string.Format(CultureInfo.InvariantCulture, message, args)));
+            //TODO leaving this in for convenience; remove before release?
```
Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs]: Can you rephrase this to "// TODO: Determine the value this tracing adds" or something more as to what the TODO is for?

**Comment by twsouthwick** [comment]
There's an `#if false` in `TypeDescriptor.cs` at line 519. Do we need to keep that?

File: src/System.ComponentModel.TypeConverter/src/Resources/Strings.resx [hunk 3]
```diff
@@ -138,6 +138,24 @@
   <data name="InvalidMemberName" xml:space="preserve">
     <value>Invalid member name</value>
   </data>
+  <data name="TypeDescriptorExpectedElementType" xml:space="preserve">
+    <value>Expected types in the collection to be of type {0}.</value>
+  </data>
+  <data name="TypeDescriptorSameAssociation" xml:space="preserve">
+    <value>Cannot create an association when the primary and secondary objects are the same.</value>
+  </data>
+  <data name="TypeDescriptorAlreadyAssociated" xml:space="preserve">
```
Comment (twsouthwick) [inline src/System.ComponentModel.TypeConverter/src/Resources/Strings.resx]: Make sure these strings are in alphabetical order

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptionProvider.cs (hunk 1 above)
Comment (chlowell) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptionProvider.cs]: I fixed the rest of these throughout the project.

**Comment by chlowell** [comment]
I changed the ```#if false``` to a feature flag documenting why it continues to exist (needs ```ReflectPropertyDescriptor```).

**Comment by twsouthwick** [comment]
LGTM
